crates/file/src/file_reader.rs

```diff
@@ -0,0 +1,37 @@
+use std::rc::Rc;
+
+use futures::Future;
+use wasm_bindgen::JsCast;
+
+use crate::blob::Blob;
+
+pub struct FileReader {
+    inner: web_sys::FileReader,
+}
+
+impl FileReader {
+    pub fn new() -> FileReader {
+        FileReader {
+            inner: web_sys::FileReader::new().unwrap(),
+        }
+    }
+
+    pub fn read_as_string(
+        self,
+        blob: &impl Blob,
+    ) -> impl futures::Future<Item = String, Error = ()> {
```

Comment:
Rather than using `impl Future` here, we should define a proper `Future` type. While using `-> impl Trait` is often convenient for the author, it is less convenient for users (since they can't use the type as a field of a struct, often aren't `Debug` or `Clone`).

Also, the struct is the thing that should contain the `Closure`, to get the lifetimes right.

```rust
#[derive(Debug)] // will actually have to be a manual impl until wasm-bindgen#1387 is fixed
pub struct ReadAsString {
    receiver: oneshot::Receiver<String>,
    closure: Closure<FnMut(...)>,
}

impl Future for ReadAsString { ... }
```

Reply:
Agree 100%.